cli/util/uploader.go

```diff
@@ -0,0 +1,429 @@
+	if err != nil {
+		return fmt.Errorf("search of index files failed: %w", err)
+	}
+
+	existingIndexCount := uint(len(objectIDs))
+	expectedIndexCount := currentHeight / indexFileSize
+
+	if existingIndexCount >= expectedIndexCount {
+		fmt.Fprintf(ctx.App.Writer, "Index files are up to date. Existing: %d, expected: %d\n", existingIndexCount, expectedIndexCount)
+		return nil
+	}
+
+	var (
+		errCh                 = make(chan error)
+		buffer                = make([]byte, indexFileSize*oidSize)
+		oidCh                 = make(chan oid.ID, indexFileSize)
+		oidFetcherToProcessor = make(chan struct{}, indexFileSize)
+	)
+	defer close(oidCh)
+	for range maxParallelSearches {
+		go func() {
+			for currOid := range oidCh {
+				err := retry(func() error {
+					obj, err := p.ObjectHead(context.Background(), containerID, currOid, signer, client.PrmObjectHead{})
+					if err != nil {
+						return fmt.Errorf("failed to fetch object %s: %w", currOid.String(), err)
+					}
+
+					blockIndex, err := getBlockIndex(*obj, blockAttributeKey)
+					if err != nil {
+						return fmt.Errorf("failed to get block index from object %s: %w", currOid.String(), err)
+					}
+					offset := (uint(blockIndex) % indexFileSize) * oidSize
+					currOid.Encode(buffer[offset:])
+					oidFetcherToProcessor <- struct{}{}
+					return nil
+				})
+				if err != nil {
+					errCh <- err
+					return
+				}
+			}
+		}()
+	}
+
+	for i := existingIndexCount + 1; i <= expectedIndexCount; i++ {
+		startIndex := (i - 1) * indexFileSize
+		endIndex := startIndex + indexFileSize
+		go func() {
+			for j := int(startIndex); j < int(endIndex); j += searchBatchSize {
+				remaining := int(endIndex) - j
+				end := j + min(searchBatchSize, remaining)
+
+				prm := client.PrmObjectSearch{}
+				filters := object.NewSearchFilters()
+				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", j), object.MatchNumGE)
+				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", end), object.MatchNumLE)
+				prm.SetFilters(filters)
+
+				objectIDs, err := neofs.ObjectSearch(ctx.Context, p, account.PrivateKey(), containerID.String(), prm)
+				if err != nil {
+					errCh <- fmt.Errorf("no OIDs found for index files: %w", err)
+					return
+				}
+
+				for _, currOid := range objectIDs {
+					oidCh <- currOid
+				}
+			}
+		}()
+
+		var completed int
+	waitLoop:
+		for {
+			select {
+			case err := <-errCh:
+				return err
+			case <-oidFetcherToProcessor:
+				completed++
+				if completed == int(indexFileSize) {
+					break waitLoop
+				}
+			}
+		}
+		select {
+		case err := <-errCh:
+			return err
+		default:
+		}
+		attrs := []object.Attribute{
+			*object.NewAttribute(attributeKey, strconv.Itoa(int(i))),
+			*object.NewAttribute("size", strconv.Itoa(int(indexFileSize))),
+		}
+		err = uploadObj(ctx.Context, *p, user.NewAutoIDSignerRFC6979(account.PrivateKey().PrivateKey), account.PrivateKey().GetScriptHash(), containerID, buffer, attrs, homomorphicHashingDisabled)
```

Comment:
A lot of repetative code of signer creation, move it to a single place and reuse.